internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,322 @@
+package scsi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+	"github.com/Microsoft/hcsshim/internal/security"
+	"github.com/Microsoft/hcsshim/internal/verity"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrNoAvailableLocation indicates that a new SCSI attachment failed because
+	// no new slots were available.
+	ErrNoAvailableLocation = errors.New("no available location")
+	// ErrNotInitialized is returned when a method is invoked on a nil [Manager].
+	ErrNotInitialized = errors.New("SCSI manager not initialized")
+	// ErrAlreadyReleased is returned when [Mount.Release] is called on a Mount
+	// that had already been released.
+	ErrAlreadyReleased = errors.New("mount was already released")
+)
+
+// Manager is the primary entrypoint for managing SCSI devices on a VM.
+// It tracks the state of what devices have been attached to the VM, and
+// mounted inside the guest OS.
+type Manager struct {
+	attachManager *attachManager
+	mountManager  *mountManager
+}
+
+// Slot represents a single SCSI slot, consisting of a controller and LUN.
+type Slot struct {
+	Controller uint
+	LUN        uint
+}
+
+// NewManager creates a new Manager using the provided backend [Attacher],
+// [Mounter], and [Unplugger], as well as other configuration parameters.
+//
+// guestMountFmt is the format string to use for mounts of SCSI devices in
+// the guest OS. It should have a single %d format parameter.
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
+	numControllers int,
+	numLUNsPerController int,
+	guestMountFmt string,
+	reservedSlots []Slot,
+) *Manager {
+	am := newAttachManager(attacher, unplugger, numControllers, numLUNsPerController, reservedSlots)
+	mm := newMountManager(mounter, guestMountFmt)
+	return &Manager{am, mm}
+}
+
+// MountConfig specifies the options to apply for mounting a SCSI device in
+// the guest OS.
+type MountConfig struct {
+	Encrypted bool
+	Options   []string
+}
+
+// Mount represents a SCSI device that has been attached to a VM, and potentially
+// also mounted into the guest OS.
```

Comment:
I agree it's confusing that internally we use "attachment" and "mount" for different things, and in the external interface glom them together as a "mount". If you have an idea for a better name I'd love to make this clearer.

Reply:
Hmm, naming is hard haha. Maybe if/when we expose the difference to the external callers in the future, we could clear this up then.

Reply:
I think part of this issue is also related to my other comment about using `MountConfig` to indicate if we should mount the attached device. From the external caller's perspective, as you mentioned, we've glommed together those two concepts into a single scsi "Mount" and are essentially hiding the fact that we're doing two separate sets of operations under the hood. But then we're having the external caller configure whether or not we do one of those sets of operations, the mounting. And we're overloading the term "mount" between the returned "Mount" object and the supplied "MountConfig".

Reply:
> My long-term hope is we can decouple "mounting" completely from the SCSI package, since it's not really a SCSI-specific operation at all.

What are you including in this decoupling? Like are you referring just to the Mounter implementations or other things as well?

Reply:
I'm not talking about a specific code change. Rather that it would make more sense for anything that deals with "take this block device in the guest and mount it" to not be tied to SCSI, since there's no real relation there (you could just as easily mount a VPMEM device, or a block file, with the same code).